### PR TITLE
isContiguous problems

### PR DESCRIPTION
**Pull request by ssnl**
Reopened version of #3091 .

cc: @apaszke

### CODE REVIEW

**Review by colesbury — DISMISSED** [review]
I'm very concerned with changing the newContiguous behavior without a very good reason. If the cuDNN API can't handle certain strides, that should be handled in the bindings to cuDNN.

**Comment by SsnL** [comment]
@soumith , what's your opinion on this?

**Comment by apaszke** [comment]
I'm fine with changing it, such that the "simple" criterion for contiguity passes, as long as we keep CUDA and CPU consistent. There's no reason why we have to limit this to cuDNN bindings if we can get it for free, without breaking compatibility in any way

**Comment by colesbury** [comment]
Except that this will break things. Currently we have the behavior that if x.is_contiguous() is True then x.contiguous() returns x. This is important: it means that you can avoid calls to contiguous if you know the tensor is already contiguous, which avoids allocation without changing behavior.

**Comment by apaszke** [comment]
Fair enough. Let's special case this in cuDNN bindings

**Comment by SsnL** [comment]
Is there a centralized place for cudnn bindings? I can't seem to find it.

**Comment by apaszke** [comment]
No, functions called `assertX` should never have side effects. We should probably patch the strides when creating descriptors

**Comment by SsnL** [comment]
Yeah I agree. I was thinking about renaming _THVoidTensor_assertContiguous as well. But patching when creating descriptors seems a better idea. I'll try that.

**Comment by SsnL** [comment]
Updated! Let me know if there is any issue :)

File: test/test_nn.py [hunk 1]
```diff
@@ -1075,6 +1075,18 @@ def test_AvgPool3d_backward_after_cat_dim1_cuda(self):
...
 
+    @unittest.skipIf(not TEST_CUDNN, "needs cudnn")
+    def test_contig_wrong_stride_cudnn(self):
+        # x has to have batch_size 1 to test contiguous checks
+        x = torch.randn(1, 16, 5, 5).cuda()
+        stride = list(x.stride())
+        stride[0] = 20
+        # change the stride in dimension 0. the tensor is still contiguous because size[0] is 1
+        x.set_(x.storage(), 0, x.size(), stride)
+        assert x.is_contiguous()
```
Comment (apaszke) [inline test/test_nn.py]: `self.assertTrue`

**Review by apaszke — COMMENTED** [review]
LGTM. Just please don't use `assert` in tests

**Review by apaszke — APPROVED** [review]

File: torch/csrc/cudnn/Descriptors.h [hunk 2]
```diff
@@ -29,6 +29,18 @@ inline cudnnDataType_t getDataType(const at::Tensor& t) {
   throw std::runtime_error("TensorDescriptor only supports double, float and half tensors");
 }
 
+static inline void fixSizeOneDimStride(int dim, int *size, int *stride) {
```
Comment (ezyang) [inline torch/csrc/cudnn/Descriptors.h]: For future reference, this is the kind of function which really deserves a comment. (My rewrite of cudnn library will add the comment so you don't have to do anything here.)

**Pull request merged**